FSW/src/util/faultManager.cpp

```diff
@@ -0,0 +1,303 @@
+/* faultManager.cpp handles NS2 payload fault mitigation
+ * Usage:
+ *  module functionality
+ *  function definitions
+ *  put function declarations in memUtil.hpp
+ * 
+ * Modules encompassed:
+ *  Fault Detection
+ *  Fault Logging
+ *  Fault Mitigation
+ *
+ * Additional files needed for compilation:
+ *  config.hpp
+ *  encodedFile.hpp
+ *  
+ */
+
+/* - - - - - - Includes - - - - - - */
+// All libraries are put in dataCollection.hpp
+// NS2 headers
+#include "../headers/faultManager.hpp"
+#include "../headers/encodedFile.hpp"
+#include "../headers/commandHandling.hpp"
+#include "../headers/housekeeping.hpp"
+
+/* Module Variable Definitions */
+// fault log
+static FaultReport faultLog[faultCode::COUNT];
+// decoded size of data on EEPROM, in terms of bytes
+const size_t EEPROM_DECODED_SIZE = faultCode::COUNT * FaultReport::MEMSIZE + PayloadData::MEMSIZE; 
+// flag to indicate whether any faults were logged in the current main loop iteration
+static bool detectedFault = false; 
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - logFault - - - - - - *
+ * Usage:
+ *  records a new occurence of a fault
+ *  
+ * Inputs:
+ *  code - fault code to log
+ * 
+ * Outputs:
+ *  None
+ */
+void logFault(uint8_t code) {
+    detectedFault = true;
+
+    // check if fault is valid
+    if (code >= faultCode::ERR_CODE) {
+        Serial.println("Warning, tried to log a fault with invalid fault code (Fault Manager)");
+        return;
+    }
+
+    Serial.print("Fault Logged: ");
+    Serial.println(code);
+
+    // update corresponding fault report
+    faultLog[code].startNum = payloadData.startCount;
+    faultLog[code].timestamp = millis();
+
+    if (faultLog[code].occurences < 255) {
+        faultLog[code].occurences++;
+        
+        // mark for action on first and max occurences
+        if (faultLog[code].occurences == 1 || faultLog[code].occurences == 255) {
```

Comment:
If taking corrective action on the first occurrence doesn't resolve the issue, it will try one more time to correct it before the fault report fills up, it's definitely redundant, but I don't think there's any harm in a little redundancy.

Reply:
agreed, that makes sense

Reply:
I think it would make more sense to try and take corrective action a few more times, then switch the spacecraft to safe mode. If software _should_ be able to correct a fault, and it fails something like 5 times in a row - something else is definitely wrong, and the ground controller should step in.

Reply:
Oooo I like that a lot. I'll save that for a future PR, once we work out some of the inevitable bugs.